### PR TITLE
Fix to #38126

### PR DESCRIPTION
**Pull request by jczaja**
### PR types
Bug fixes

### PR changes
OPs

### Describe
This PR is fixing #38126 . In particular it fixes braodcasting of elementwise operation of oneDNN kernels for NHWC scenario.

### CODE REVIEW

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 1]
```diff
@@ -643,11 +643,21 @@ class BinaryMKLDNNHandler
       std::vector<int64_t> dims1_ex(rankdiff, 1);
       dims1_ex.insert(next(dims1_ex.begin(), (axis == -1 ? rankdiff : axis)),
                       src_y_tz.begin(), src_y_tz.end());
+      // For broadcasting for NHWC we need rotate extended shape
+      if (MKLDNNDeviceContext::tls().get_cur_paddle_data_layout() ==
+          framework::DataLayout::kNHWC) {
+        std::rotate(dims1_ex.begin() + 1, dims1_ex.end() - 1, dims1_ex.end());
+      }
```
Comment (chen543345) [inline paddle/fluid/platform/mkldnn_reuse.h:650]: 优秀

Comment (chen543345) [reply paddle/fluid/platform/mkldnn_reuse.h:650]: 很好

Comment (zpf1222) [reply paddle/fluid/platform/mkldnn_reuse.h:650]: good

Comment (zpf1222) [reply paddle/fluid/platform/mkldnn_reuse.h:650]: > 很好



> 很好

aaa

**Comment by lidanqing-vv** [comment]
@baoachun Please review

File: paddle/fluid/operators/activation_op.cc [hunk 2]
```diff
@@ -128,6 +128,26 @@ class ActivationOp : public framework::OperatorWithKernel {
       const framework::ExecutionContext& ctx) const override {
     return GetKernelType(ctx, *this, "X");
   }
+
+  framework::OpKernelType GetKernelTypeForVar(
+      const std::string& var_name, const Tensor& tensor,
+      const framework::OpKernelType& expected_kernel_type) const {
+#ifdef PADDLE_WITH_MKLDNN
+    // When activation is first oneDNN op (there was some non oneDNN op
```
Comment (baoachun) [inline paddle/fluid/operators/activation_op.cc:136]: Do all operators need to adapt to the format of the input NHWC when it is the first mkldnn operator?

Comment (jczaja) [reply paddle/fluid/operators/activation_op.cc:136]: Yes. So this PR is only doing this change for fixing #38126 , but in general similar change has to be added to other oneDNN ops.

**Comment by jczaja** [comment]
@baoachun  I have a problem with CI :+1: 
../paddle/pten/core/sparse_coo_tensor.h(58): **error: function returning abstract class "pten::SparseCooTensor" is not allowed:**
This PR is not touching this pten, so please advice how to proceed to have it resolved

**Comment by baoachun** [comment]
Hi @jczaja , it may be the compilation failure caused by other people's pr. I have triggered PR-CI-Build. If it still fails, I will contact my colleagues to help solve it.

**Comment by jczaja** [comment]
@Silv3S , @piotrekobiIntel  Please help with review

File: paddle/fluid/operators/activation_op.cc (hunk 2 above)
Comment (ghost) [reply paddle/fluid/operators/activation_op.cc:136]: Can this be added to a class that other operators inherit?

**Review by ghost — APPROVED** [review]

**Comment by Silv3S** [comment]
LGTM

**Comment by jczaja** [comment]
@baoachun Please continue your review

**Comment by jczaja** [comment]
@piotrekobiIntel Thanks for suggesstion. Technically yes but in practice we need to create another of class to fit this code there. Code is a bit diffrent for operators with parameters and may differ even for parameter-less ops due to some input and output vars having diffrent names. So in fact adding another class sounds like significant  architectural effort while at least half of ops will override this code anyway due to diffrences. So for now I will leave it like this and when more ops got similar code then We can judge if it is profitable to make a separate class with this code.

**Review by baoachun — APPROVED** [review]
LGTM

**Review by lidanqing-vv — APPROVED** [review]
LGTM

**Pull request merged**